diff -ur CoinMP-1.7.0/CoinMP/src/CoinProblem.c CoinMP-1.7.0-new/CoinMP/src/CoinProblem.c
--- CoinMP-1.7.0/CoinMP/src/CoinProblem.c	2011-11-02 23:43:29.000000000 -0400
+++ CoinMP-1.7.0-new/CoinMP/src/CoinProblem.c	2013-11-12 11:26:05.167841333 -0500
@@ -363,7 +363,7 @@
 	double RangeABS, RangeValue;
 
 	if (pProblem->RowCount == 0) {
-		return 0;
+		return 1;
 	}
 	pProblem->RowLower = (double* )malloc(pProblem->RowCount*sizeof(double));
 	pProblem->RowUpper = (double* )malloc(pProblem->RowCount*sizeof(double));
